src/io/fastx.rs

```diff
@@ -0,0 +1,319 @@
+    FASTQ(fastq::Record),
+}
+
+impl From<fasta::Record> for EitherRecord {
+    fn from(record: fasta::Record) -> Self {
+       EitherRecord::FASTA(record) 
+    }
+}
+
+impl From<fastq::Record> for EitherRecord {
+    fn from(record: fastq::Record) -> Self {
+       EitherRecord::FASTQ(record) 
+    }
+}
+
+impl Record for EitherRecord {
+    matchthrough!(is_empty, bool);
+    matchthrough!(check, Result<(), &str>);
+    matchthrough!(id, &str);
+    matchthrough!(desc, Option<&str>);
+    matchthrough!(seq, TextSlice<'_>);
+}
+
+enum EitherRecords<R: io::Read> {
+    FASTA(fasta::Records<R>),
+    FASTQ(fastq::Records<R>),
+}
+
+pub struct Records<R: io::Read> {
+    // This isn't as type safe as I'd like it to be
+    records: Option<EitherRecords<R>>,
+    reader: Option<R>,
+}
+
+impl<R: io::Read> Iterator for Records<R> {
+    type Item = io::Result<EitherRecord>;
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(reader) = mem::replace(&mut self.reader, None) {
+            let mut reader: io::BufReader<R> = io::BufReader::new(reader);
+            let mut line = String::new();
+            if let Err(e) = reader.read_line(&mut line) {
+                return Some(Err(e));
+            }
+            match line.chars().next() {
+                Some('>') => self.records = Some(EitherRecords::FASTA(fasta::Reader::new_with_line(reader, line).records())),
+                Some('@') => self.records = Some(EitherRecords::FASTQ(fastq::Reader::new_with_line_buffer(reader, line).records())),
+                Some(c) => return Some(Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("File is not a valid FASTA/FASTQ, illegal start character '{}'", c),
+                ))),
+                None => return None,
+            }
+        }
+        match &mut self.records {
+            Some(EitherRecords::FASTA(r)) => r
+                .next()
+                .map(|record_res| record_res.map(EitherRecord::from)),
+            Some(EitherRecords::FASTQ(r)) => r
+                .next()
+                .map(|record_res| record_res.map(EitherRecord::from)),
+             None => None,
+        }
+    }
+}
+
+impl<R: io::Read> From<R> for Records<R> {
+    fn from(reader: R) -> Self {
+        Records { records: None, reader: Some(reader) }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum FastxType {
+    Fastq,
+    Fasta,
+}
+
+pub fn get_type<R: io::Read + io::Seek>(reader: &mut R) -> Result<FastxType, io::Error> {
+    let mut first = [0];
+    reader.read_exact(&mut first)?;
+    reader.seek(SeekFrom::Current(-1))?;
+
+    match first[0] as char {
+        '>' => Ok(FastxType::Fasta),
+        '@' => Ok(FastxType::Fastq),
+        _ => {
+            let msg = format!("first character was '{}' which is not a valid first character for a fastq or fasta file", first[0]);
+            Err(io::Error::new(io::ErrorKind::InvalidData, msg))
+        }
+    }
+}
```

Comment:
I think we can remove the `io::Seek` constraint by use something similar as what is done in niffler https://github.com/luizirber/niffler/blob/master/src/basic/mod.rs#L37

We can maybe have two function:
- `get_type` use niffler tricks
- `get_type_seek` use what you already write.

Reply:
Makes sense. I was also going to include one that takes a path for convenience. Thanks for showing me that trick! I was trying to create something like that manually instead of using the `new_with_line` style approach but my implementation was over-complicated. I may switch to this in the reader as well for simplicity.